### PR TITLE
adding subtomograms extraction

### PR DESCRIPTION
**Pull request by Vilax**

### CODE REVIEW

**Comment by pconesa** [comment]
One question, since I guess this needs some xmipp code....this cannot be released unless xmipp is released.

Right?

**Comment by fede-pe** [comment]
This PR requires approving first https://github.com/I2PC/xmipp/pull/691

**Comment by Vilax** [comment]
Yes, but xmipp is periodically released, so it should not be a problem. For now it will be in devel

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+from tomo.protocols import ProtTomoBase
+from xmipp3.convert import alignmentToRow
+
+
+COORD_BASE_FN = 'coords'
+# Tomogram type constants for particle extraction
+SAME_AS_PICKING = 0
+OTHER = 1
+
+
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:51]: would be nice to define possible outputs --> https://github.com/scipion-em/scipion-pyworkflow/blob/devel/pyworkflow/protocol/protocol.py#L332

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    #_possibleOutputs = SetOfSubTomograms
+    #OUTPUT_PREFIX = _possibleOutputs.subtomograms.name
+    lines = []
+    tomoFiles = []
+
+    # --------------------------- DEFINE param functions ------------------------
+    def _defineParams(self, form):
+        form.addSection(label='Parameters')
+
+        form.addParam('tomograms', PointerParam, pointerClass=SetOfTomograms,
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:67]: Since coordinates have a set of tomograms associated I'd ask for 3d coordinates FIRST, , and optionally, a setOfTomograms

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                      label='Tomograms', help='The subtomograms will be extracted from this set.')
+
+        form.addParam('coords', PointerParam, pointerClass=SetOfCoordinates3D,
+                  label='Coordinates', help='The coordinates that definethe position of the set of subtomograms to be extracted.'
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:71]: I'd rephrase it: "3D coordinates to use in the extraction process". ?

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                      help='Normally, tomograms has the contrast inverted with respect to standard for subtomograms. '
+                           'It means, in the tomograms the structure is black and the noise is white. Generally, the'
+                           'subtomograms are white with a black background. This means that the subtomograms has the '
+                           'contrast inverted with respect to the tomograms. Put this flag as True if the contrast'
+                           'need to be inverted.')
+
+    # --------------------------- INSERT steps functions ------------------------
+    def _insertAllSteps(self):
+
+        for tom in self.tomograms.get():
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:90]: I think this is better to get only the tomograms present in the coordinates in case there is a subset of 3d coordinates.

Use --> https://github.com/scipion-em/scipion-em-tomo/blob/devel/tomo/objects.py#L1501

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            tomId = tom.getObjId()
+            self._insertFunctionStep(self.extractStep, self.tomograms.get(), tomId)
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:92]: Why passing the set.....just pass the tomId and get the tomogram inside extractStep. The params are serialized, in this case for nothing.

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                if os.path.basename(tomoFile) == os.path.basename(item.getFileName()):
+                    coordSet = self.lines[ind]
+                    tomId = item.getObjId()
+                    tsId = self.tomograms.get()[tomId].getTsId()
+                    outputSet, counter = self.readSetOfSubTomograms(tomoFile,
+                                                                    self.outputSubTomogramsSet,
+                                                                    coordSet, 1, counter, tsId)
+
+        self._defineOutputs(outputSetOfSubtomogram=outputSet)
+        self._defineSourceRelation(self.coords, outputSet)
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:183]: good! relation quite often is forgotten Well done.

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        subtomoFileList = sorted(glob.glob(outRegex))
+
+        for idx, subtomoFile in enumerate(subtomoFileList):
+
+            self.debug("Registering subtomogram %s - %s" % (counter, subtomoFile))
+
+            subtomogram = SubTomogram()
+            subtomogram.cleanObjId()
+            subtomogram.setLocation(subtomoFile)
+            subtomogram.setCoordinate3D(coordSet[idx])
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:200]: You are asumming set of coordinates is not a sub set).

This will fail in this case... you can't use idx (I think) to identify the coordinates.

File: xmipptomo/tests/test_protocol_extract_subtomos.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                                                   filesPath=self.coords3D,
+                                                   importTomograms=protImportTomogram.outputTomograms,
+                                                   filesPattern='', boxSize=32,
+                                                   samplingRate=5)
+
+        self.launchProtocol(protImportCoordinates3d)
+        self.assertSetSize(protImportTomogram.outputTomograms, 1,
+                           "There was a problem with tomogram output")
+        self.assertSetSize(protImportCoordinates3d.outputCoordinates, 5,
+                           "There was a problem with coordinates 3d output")
```
Comment (pconesa) [inline xmipptomo/tests/test_protocol_extract_subtomos.py:77]: May be add a subset protocol in between to test the subset case.

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+            for ind, tomoFile in enumerate(self.tomoFiles):
+                if os.path.basename(tomoFile) == os.path.basename(item.getFileName()):
+                    coordSet = self.lines[ind]
+                    tomId = item.getObjId()
+                    tsId = self.tomograms.get()[tomId].getTsId()
+                    outputSet, counter = self.readSetOfSubTomograms(tomoFile,
+                                                                    self.outputSubTomogramsSet,
+                                                                    coordSet, 1, counter, tsId)
+
+        self._defineOutputs(outputSetOfSubtomogram=outputSet)
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:182]: We are avoiding those long/unfrinedly names as outputs. Better: "Subtomograms"

**Review by pconesa — CHANGES_REQUESTED** [review]
Thanks vilas, test are passing. My mayor concern is that this will not work with subsets of coordinates.

**Comment by pconesa** [comment]
Wait! Actually it works with a subset

File: xmipptomo/protocols/protocol_extract_subtomos.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            subtomogram.cleanObjId()
+            subtomogram.setLocation(subtomoFile)
+            subtomogram.setCoordinate3D(coordSet[idx])
+            transformation = coordSet[idx]._eulerMatrix
+            shift_x, shift_y, shift_z = transformation.getShifts()
+            transformation.setShifts(factor * shift_x,
+                                     factor * shift_y,
+                                     factor * shift_z)
+            subtomogram.setTransform(transformation, convention=const.TR_EMAN)
+            subtomogram.setVolName(tomoFile)
```
Comment (pconesa) [inline xmipptomo/protocols/protocol_extract_subtomos.py:207]: volname is just the basename should match the tomogram.tsId

File: xmipptomo/protocols/protocol_extract_subtomos.py (hunk 7 above)
Comment (pconesa) [reply xmipptomo/protocols/protocol_extract_subtomos.py:200]: well, is not failing because of the subset, but your subtomoFileList is sorted by name so:

1
10
11
2
3
...

whereas coordinates do not match this same order.

**Review by fede-pe — APPROVED** [review]
Test passing in my local installation. Ready to merge.
